### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by WaffleLapkin**
- Remove `derive(DialogueState)` macro
- Remove `get_or_return!`, use `?`/`Result` instead
- Don't use `quote::__private`
- Improve error handling by adding custom `Error` type
- Mark top-level items as `pub(crate)`
- Cleanup renaming
- Clean up attribute handling a bit
- fix typos & such
- Refactor attributes some more
- Iterator ftw

### CODE REVIEW

File: src/attr.rs [hunk 1]
```diff
@@ -1,68 +1,96 @@
...
-            "description" => Ok(BotCommandAttribute::Description),
-            "rename" => Ok(BotCommandAttribute::RenameRule),
-            "parse_with" => Ok(BotCommandAttribute::CustomParser),
-            "separator" => Ok(BotCommandAttribute::Separator),
-            _ => Err(syn::Error::new(name_arg.span(), "unexpected argument")),
+            "prefix" => Ok(CommandAttrName::Prefix),
+            "description" => Ok(CommandAttrName::Description),
+            "rename" => Ok(CommandAttrName::Rename),
+            "parse_with" => Ok(CommandAttrName::ParseWith),
+            "separator" => Ok(CommandAttrName::Separator),
```
Comment (Hirrolot) [inline src/attr.rs:25]: Maybe it's better to add `use CommandAttrName::*;` prior to `match`?

**Review by Hirrolot — CHANGES_REQUESTED** [review]
LGTM overall, with some minor issues (I've left some comments). Before merging, we should test it in a separate branch in teloxide, since it does also contain some macro tests.

File: src/attr.rs (hunk 1 above)
Comment (WaffleLapkin) [reply src/attr.rs:25]: Makes sense 👍🏻 

Maybe we (you) should add this recommendation to style guide?

**Review by WaffleLapkin — COMMENTED** [review]
I think this refactor is mostly done, this PR can be reviewed/merged.

Future work:
- Maybe refactor `CommandEnum`/`Command`, they seem to be duplicating a lot
- Fix the `FIXME`s
- Do the changes proposed in https://github.com/teloxide/teloxide/issues/635
- Write documentation
  - For internal parts such as `fold_attrs`, so that code is easier to maintain/read/understand
  - For external parts such that docs of the crate are not empty and they refer to teloxide

File: src/attr.rs (hunk 1 above)
Comment (Hirrolot) [reply src/attr.rs]: Ok, will add 👍🏻

**Review by Hirrolot — CHANGES_REQUESTED** [review]
`CHANGELOG.md` :)

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 ## unreleased
 
+### Removed
+
+- Remove `derive(DialogueState)` macro
+
+### Changed
+
+- `#[command(rename = "a_name_that_is_not_a_case_name")]` doesn't work anymore
+- A lot of internal refactoring
```
Comment (Hirrolot) [inline CHANGELOG.md:16]: Not sure that it needs to be in the changelog since "refactoring" doesn't reflect any API changes.

**Review by Hirrolot — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (Hirrolot) [inline CHANGELOG.md:16]:
```suggestion
```

**Review by Hirrolot — CHANGES_REQUESTED** [review]

**Review by Hirrolot — APPROVED** [review]

**Pull request merged**